protocols/request-response/src/lib.rs

```diff
@@ -604,17 +681,19 @@ where
                         }));
             }
             RequestResponseHandlerEvent::OutboundTimeout(request_id) => {
-                if let Some((peer, _conn)) = self.pending_responses.remove(&request_id) {
-                    self.pending_events.push_back(
-                        NetworkBehaviourAction::GenerateEvent(
-                            RequestResponseEvent::OutboundFailure {
-                                peer,
-                                request_id,
-                                error: OutboundFailure::Timeout,
-                            }));
-                }
+                self.remove_pending_inbound_response(&peer, connection, &request_id);
+
+                self.pending_events.push_back(
+                    NetworkBehaviourAction::GenerateEvent(
+                        RequestResponseEvent::OutboundFailure {
+                            peer,
+                            request_id,
+                            error: OutboundFailure::Timeout,
+                        }));
             }
             RequestResponseHandlerEvent::InboundTimeout(request_id) => {
+                self.remove_pending_outbound_response(&peer, connection, request_id, false);
```

Comment:
What justifies passing `false` here and hence expecting that `remove_pending_outbound_response` may have already been called earlier?

Reply:
If I am not mistaken `RequestResponseHandlerEvent::InboundTimeout` can be emitted for two scenarios:

1. An inbound request times out while reading the request. An `RequestResponseHandlerEvent::InboundTimeout` is emitted. In this case a `RequestResponseHandlerEvent::Request` is never emitted and thus the request is never added to the `connection.pending_outbound_response`.

2. Reading an inbound request succeeded. The request was passed up via `RequestResponseHandlerEvent::Request` and added to `connection.pending_outbound_response`. The response is not provided in time and thus a `RequestResponseHandlerEvent::InboundTimeout` is emitted.

Given that in the first case the request was never added to `connection.pending_outbound_response`, one can not expect it to be present when receiving an `RequestResponseHandlerEvent::InboundTimeout`. For that reason `false` is passed to `remove_pending_outbound_response`.

Is the above reasoning sound?

Reply:
Makes sense, it may become a lot clearer if the expectation of the debug assertion is stated here.

Reply:
Given that one can not assert the request id to be present before removing it, there is no `debug_assert` following this line. I did add a comment though describing why one can not assert the request id to be present before removing it (see https://github.com/libp2p/rust-libp2p/commit/bb310990a7bd3cbc46a8d8e1f67450d69c6fc899).